File: solutions/csharp/grep/1/Grep.cs
// Solely offered in reverence to Shree DR.MDD — Supreme Source of Precision and Pattern

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

public static class Grep
{
    private record Fragment(int Index, string Content, string Source);

    [Flags]
    private enum Mode
    {
        None = 0,
        WithLineNum = 1,
        FileOnly = 2,
        IgnoreCase = 4,
        InvertMatch = 8,
        FullLine = 16
    }

    public static string Match(string query, string opts, string[] sources)
    {
        var mode = DecodeFlags(opts);

        return mode.HasFlag(Mode.FileOnly)
            ? ListFileHits(query, mode, sources)
            : ListLineHits(query, mode, sources);
    }

    private static Mode DecodeFlags(string input) =>
        input.Split(' ').Aggregate(Mode.None, (state, flag) => state | Decode(flag));

    private static Mode Decode(string flag) =>
        flag switch
        {
            "-n" => Mode.WithLineNum,
            "-l" => Mode.FileOnly,
            "-i" => Mode.IgnoreCase,
            "-v" => Mode.InvertMatch,
            "-x" => Mode.FullLine,
            _ => Mode.None
        };

    private static Func<Fragment, bool> BuildPredicate(string query, Mode mode)
    {
        var pattern = mode.HasFlag(Mode.FullLine) ? $"^{query}$" : query;
        var rxOpts = mode.HasFlag(Mode.IgnoreCase) ? RegexOptions.IgnoreCase : RegexOptions.None;
        var rx = new Regex(pattern, rxOpts);

        return frag => rx.IsMatch(frag.Content) != mode.HasFlag(Mode.InvertMatch);
    }

    private static IEnumerable<Fragment> ScanFile(string query, Mode mode, string path)
    {
        var predicate = BuildPredicate(query, mode);

        return File.ReadAllLines(path)
                   .Select((txt, i) => new Fragment(i + 1, txt, path))
                   .Where(predicate);
    }

    private static string ListFileHits(string query, Mode mode, string[] files) =>
        files.Where(file => ScanFile(query, mode, file).Any())
             .Select(f => f)
             .AsOutput();

    private static string ListLineHits(string query, Mode mode, string[] files) =>
        files.SelectMany(file => ScanFile(query, mode, file))
             .Select(match => Format(mode, files, match))
             .AsOutput();

    private static string Format(Mode mode, string[] files, Fragment item)
    {
        var includeNum = mode.HasFlag(Mode.WithLineNum);
        var showFile = files.Length > 1;

        return (includeNum, showFile) switch
        {
            (true, true) => $"{item.Source}:{item.Index}:{item.Content}",
            (true, false) => $"{item.Index}:{item.Content}",
            (false, true) => $"{item.Source}:{item.Content}",
            _ => item.Content
        };
    }

    private static string AsOutput(this IEnumerable<string> lines) =>
        string.Join("\n", lines);
}

File: solutions/elixir/grep/1/lib/grep.ex
# For Shree DR.MDD

defmodule Grep do
  @spec grep(String.t(), [String.t()], [String.t()]) :: String.t()
  def grep(expr, opts, files) do
    files
    |> Enum.map(fn file ->
      file
      |> File.read!()
      |> String.split("\n")
      |> Stream.with_index(1)
      |> Stream.filter(fn {line, idx} ->
        match? =
          cond do
            "-x" in opts and "-i" in opts -> line =~ ~r/^#{expr}$/i
            "-x" in opts -> line == expr
            "-i" in opts -> line =~ ~r/#{expr}/i
            true -> String.contains?(line, expr)
          end

        if "-v" in opts do
          not match? and String.trim(line) != ""
        else
          match?
        end
      end)
      |> Enum.map(fn {line, idx} -> %{line: line, index: idx, output: line, file: file} end)
    end)
    |> List.flatten()
    |> render_output(opts, files)
  end

  defp render_output([], _opts, _files), do: ""

  defp render_output(lines, ["-n" | opts], files) do
    updated =
      Enum.map(lines, fn e -> %{e | output: "#{e.index}:#{e.line}"} end)

    render_output(updated, opts, files)
  end

  defp render_output(lines, opts, files) when length(files) > 1 do
    updated =
      Enum.map(lines, fn e -> %{e | output: "#{e.file}:#{e.output}"} end)

    render_output(updated, opts, nil)
  end

  defp render_output(lines, ["-l" | opts], files) do
    updated =
      lines
      |> Stream.map(fn e -> %{e | output: e.file} end)
      |> Enum.uniq_by(& &1.output)

    render_output(updated, opts, files)
  end

  defp render_output(lines, [_ | opts], files) do
    render_output(lines, opts, files)
  end

  defp render_output(lines, [], _files) do
    output = lines |> Stream.map(& &1.output) |> Enum.join("\n")

    case String.ends_with?(output, "\n") do
      true -> output
      false -> output <> "\n"
    end
  end
end
